### PR TITLE
Fix conversion to float32 dtype

### LINKED ISSUE #4189: skimage.util.convert not converting to 'float32' with numpy v1.13.3
## Description

During my explorations for PR #4161 CI failure, I found that `skimage.util.convert` doesn't convert double precision images to single precision with `numpy v1.13.3` when called with `dtype='float32'`. Even more strange, it works when `dtype=np.float32`

## Way to reproduce

```python
>>> import numpy as np
>>> from skimage.util.dtype import convert
>>> a = np.random.rand(5, 5)
>>> convert(a, 'float32').dtype
dtype('float64')
>>> convert(a, np.float32).dtype
dtype('float32')
>>> np.__version__
'1.13.3'
```

This doesn't happen with more recent numpy version:

```python
>>> a = np.random.rand(5, 5)
>>> convert(a, 'float32').dtype
dtype('float32')
>>> convert(a, np.float32).dtype
dtype('float32')
>>> np.__version__
'1.16.4'
```

## Version information

```python
3.6.7 | packaged by conda-forge | (default, Jul  2 2019, 02:18:42) 
[GCC 7.3.0]
Linux-4.15.0-64-generic-x86_64-with-debian-buster-sid
scikit-image version: 0.16.dev0
numpy version: 1.13.3

```

### PR DESCRIPTION
**Pull request by rfezzani**
## Description

Closes #4189.

This PR consists in:

- refactoring `skimage.util.convert`: defining annex functions (`_scale`, `_dtype_bits`, `_dtype_itemsize`) outside the body of the `convert` function to make it more readable,
- refactoring of `_dtype_bits`: removing the definition of `compare` function,
- using `np.obj2sctype(dtype)` instead of `type_out` to fix the conversion to `float32` whatever is `numpy` version,
- adding tests for conversion from any combination of floating data type definition (`float32`, np.float, float, etc...)
- adding tests for conversion from a subclass of the desired output datatype (check for no regression).

## Checklist

- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [ ] Gallery example in `./doc/examples` (new features only)
- [ ] Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- [x] Unit tests
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

## For reviewers

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.
- [ ] Consider backporting the PR with `@meeseeksdev backport to v0.14.x`

### CODE REVIEW

**Comment by sciunto** [comment]
I'm +1 with @hmaarrfk : bump numpy + keep the tests.

**Comment by rfezzani** [comment]
I removed inner functions because they are not required by design and they make `convert` function body length from 262 lines to 173. It made my debugging much easier!
I would also refer to @jni [comment](https://github.com/scikit-image/scikit-image/pull/4175#issuecomment-533756432) last item concerning functions length...
But well, if you all agree and request reverting this refactoring, I will do it ;-)

**Comment by sciunto** [comment]
It's not really the fact you separated the function but rather the shortcuts such as `compare_bits_to = bits.__le__ if kind == 'u' else bits.__lt__` that make me less happy with the modifications.

File: skimage/util/dtype.py [hunk 1]
```diff
@@ -54,6 +54,126 @@ def dtype_limits(image, clip_negative=False):
...
+        The data type object element size.
+
+    Returns
+    -------
+    dtype: data type object
+        Data type of `kind` that can store a `bits` wide unsigned int
+
+    """
+
+    compare_bits_to = bits.__le__ if kind == 'u' else bits.__lt__
```
Comment (jni) [inline skimage/util/dtype.py]: I agree with @sciunto, undo this particular optimisation, as it is much less readable than the inner `compare` function. Something preferable would be to just be explicit in the loop:

```python
s = next(i for i in (itemsize,) + (2, 4, 8)
         if bits < i*8 or bits == i*8 and kind == 'u')
```

Comment (rfezzani) [reply skimage/util/dtype.py]: I like this ;-) well done @jni

**Comment by rfezzani** [comment]
Thx @sciunto, I am still stuck with Travis failures :-( Even using `  # doctest: +NORMALIZE_WHITESPACE` doesn't solve the problem :disappointed:

**Comment by sciunto** [comment]
Try numpy 0.14.1. Looks like there is a fix backported in this version that might be related to this.

**Comment by sciunto** [comment]
It works.

File: skimage/measure/_moments.py [hunk 2]
```diff
@@ -94,7 +94,8 @@ def moments_coords_central(coords, center=None, order=3):
     point, this no longer holds:
 
     >>> coords2 = np.concatenate((coords, [[17, 17]]), axis=0)
-    >>> np.round(moments_coords_central(coords2), 2)
+    >>> np.round(moments_coords_central(coords2),
+    ...          decimals=2)  # doctest: +NORMALIZE_WHITESPACE +ELLIPSIS
```
Comment (sciunto) [inline skimage/measure/_moments.py]: I don't think ELLIPSIS is need. The output does not contain `...`

**Comment by jni** [comment]
@hmaarrfk this is ready for you to pull the trigger! =) Thanks all!

**Comment by hmaarrfk** [comment]
didn't know that my change request was blocking. I rarely mean to have blocking requests..

**Comment by jni** [comment]
@hmaarrfk typically when I have requests that I don't want to block, I use "comment" rather than "request changes". Otherwise the big red x remains. I can "dismiss" it, but that feels, well, dismissive. =D

**Pull request merged**